src/distilabel/pipeline.py

```diff
@@ -611,26 +612,51 @@ def _generate(  # noqa: C901
                         labels.append(batch_labels)  # type: ignore
                     else:
                         labels.extend(batch_labels)  # type: ignore
+
                 except Exception as e:
-                    if not enable_checkpoints:
+                    if not checkpoint_strategy:
                         raise RuntimeError(
-                            "`Pipeline.generate` failed during labelling step. Setting `enable_checkpoints=True` is recommended!"
+                            "`Pipeline.generate` failed during labelling step. Passing a `DatasetCheckpoint` is recommended!"
                         ) from e
                     logger.error(
                         f"`Pipeline.generate` failed during labelling step with exception: {e}"
                     )
+
                     return self._build_dataset(
                         dataset,
                         generations=generations,
                         labels=labels,
                         batch_size=batch_size,
                     )
 
+            if checkpoint_strategy and checkpoint_strategy.do_checkpoint(
+                batch_i * batch_size
+            ):
+                logger.info(f"Saving dataset up to batch {batch_i}...")
+                ds = self._build_dataset(
+                    dataset,
+                    generations=generations,
+                    labels=labels,
+                    batch_size=batch_size,
+                )
+                ds.save_to_disk(
+                    checkpoint_strategy.path,
+                    **checkpoint_strategy.extra_kwargs,
+                )
+
         _pipeline_progress.stop()
 
-        return self._build_dataset(
+        ds = self._build_dataset(
             dataset, generations=generations, labels=labels, batch_size=batch_size
         )
+        if checkpoint_strategy:
+            logger.info("Saving final dataset...")
+            ds.save_to_disk(
+                checkpoint_strategy.path,
+                **checkpoint_strategy.extra_kwargs,
+            )
+
```

Comment:
```suggestion
        if checkpoint_strategy:
            ds.save_to_disk(
                checkpoint_strategy.path,
                **checkpoint_strategy.extra_kwargs,
            )
            logger.info("Final dataset saved.")

```

Reply:
Will update this one with the checkpoint path as suggested in the previous review 😄